<a href="https://colab.research.google.com/github/TrishaBalamurugan/AU---Projects/blob/main/Assignment_4_Comp_5630.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Housing Classification – Naïve Bayes & Decision Tree

## Overview
In this assignment, I built two machine learning models to classify houses into:
- Apartment  
- House  
- Condo  

The models used are:
1. Naïve Bayes  
2. Decision Tree  

The dataset was provided in an Excel file with separate training and test sets.

---

## Dataset
File: `Asssignment4_Data.xlsx`

Features include:
- Bathrooms  
- Bedrooms  
- Garages  
- Age of home  
- Local Price  
- (and others)

Target variable: **Construction type**

---

##  Naïve Bayes Part 1 (Hand Calculations)

The detailed hand calculations for the conditional probability distributions are included in a separate document:

**File:** NaiveBayes_Part1.pdf

This document shows the step-by-step computation of probabilities for each feature as required in the assignment.

---

## Naïve Bayes Part 2 code and output
- Calculated prior and conditional probabilities from training data  
- Used Bathrooms, Bedrooms, and Garages  
- Applied the MAP rule to classify test examples  

**Observation:**  
Some probabilities became 0 because certain feature values didn’t appear in some classes (zero-frequency issue).

---

## Decision Tree code and output
- Used Scikit-learn’s DecisionTreeClassifier  
- Default model:
  - Train Accuracy = 1.0  
  - Test Accuracy = 0.4 (overfitting)

### Depth Experiment:
- Best result at **max_depth = 2**
- Test Accuracy improved to **0.8**

**Observation:**  
Smaller depth reduces overfitting and improves performance.

---

## Conclusion
Naïve Bayes is simple and works well but has limitations with zero probabilities.  
Decision Trees perform better after tuning (like limiting depth).  
Overall, controlling model complexity is important for better results.

---

## ChatGPT Declaration
I used ChatGPT to help understand concepts, debug code, and improve explanations, but all final work reflects my own understanding.

# Naive Bayes Part 1 Hand Calculations PDF

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Naives Bayes Part 1.pdf to Naives Bayes Part 1.pdf


In [ ]:
from google.colab import files
files.download("Naives Bayes Part 1.pdf")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
from IPython.display import display, HTML

display(HTML('<a href="Naives Bayes Part 1.pdf" target="_blank">Open Naive Bayes Part 1 PDF</a>'))

#Naive Bayes Code - Part 2

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Asssignment4_Data.xlsx to Asssignment4_Data.xlsx


In [ ]:
import pandas as pd

file = "Asssignment4_Data.xlsx"

data = pd.read_excel(file, sheet_name=None)

print("Sheets in file:", data.keys())

Sheets in file: dict_keys(['Train', 'Test'])


In [ ]:
import pandas as pd

# -----------------------------
# LOAD DATA
# -----------------------------
file = "Asssignment4_Data.xlsx"

train_df = pd.read_excel(file, sheet_name="Train")
test_df = pd.read_excel(file, sheet_name="Test")

# -----------------------------
# PRIOR PROBABILITIES
# -----------------------------
P_Apartment = 8/25
P_House = 9/25
P_Condo = 8/25

# -----------------------------
# CONDITIONAL PROBABILITIES
# -----------------------------

# Bathrooms
P_bath_A = {1: 5/8, 1.5: 2/8, 2.5: 1/8}
P_bath_H = {1: 6/9, 1.5: 2/9, 2.5: 1/9}
P_bath_C = {1: 4/8, 1.5: 3/8, 2.5: 1/8}

# Bedrooms
P_bed_A = {3: 6/8, 4: 2/8, 5: 0}
P_bed_H = {3: 6/9, 4: 3/9, 5: 0}
P_bed_C = {3: 6/8, 4: 0, 5: 2/8}

# Garages
P_gar_A = {0: 2/8, 1: 5/8, 1.5: 1/8, 2: 0}
P_gar_H = {0: 0, 1: 4/9, 1.5: 0, 2: 5/9}
P_gar_C = {0: 0, 1: 6/8, 1.5: 0, 2: 2/8}


# -----------------------------
# PREDICTION FUNCTION
# -----------------------------
def predict(bath, bed, gar):

    # Apartment
    pA = P_Apartment
    pA *= P_bath_A.get(bath, 0)
    pA *= P_bed_A.get(bed, 0)
    pA *= P_gar_A.get(gar, 0)

    # House
    pH = P_House
    pH *= P_bath_H.get(bath, 0)
    pH *= P_bed_H.get(bed, 0)
    pH *= P_gar_H.get(gar, 0)

    # Condo
    pC = P_Condo
    pC *= P_bath_C.get(bath, 0)
    pC *= P_bed_C.get(bed, 0)
    pC *= P_gar_C.get(gar, 0)

    # Normalize probabilities
    total = pA + pH + pC
    if total != 0:
        pA /= total
        pH /= total
        pC /= total

    # MAP decision
    prediction = max(
        [("Apartment", pA), ("House", pH), ("Condo", pC)],
        key=lambda x: x[1]
    )[0]

    return pA, pH, pC, prediction


# -----------------------------
# RUN ON TEST DATA
# -----------------------------
print("===== Naïve Bayes Results =====")

for i, row in test_df.iterrows():

    bath = row["Bathrooms"]
    bed = row["# Bedrooms"]
    gar = row["# Garages"]

    pA, pH, pC, pred = predict(bath, bed, gar)

    print(f"\nTest Example {i+1}: ({bath}, {bed}, {gar})")
    print(f"P(Apartment | X) = {pA:.4f}")
    print(f"P(House | X) = {pH:.4f}")
    print(f"P(Condo | X) = {pC:.4f}")
    print(f"Final Prediction (MAP) = {pred}")

===== Naïve Bayes Results =====

Test Example 1: (1.5, 3, 1.0)
P(Apartment | X) = 0.2914
P(House | X) = 0.1842
P(Condo | X) = 0.5245
Final Prediction (MAP) = Condo

Test Example 2: (1.5, 4, 2.0)
P(Apartment | X) = 0.0000
P(House | X) = 1.0000
P(Condo | X) = 0.0000
Final Prediction (MAP) = House

Test Example 3: (1.0, 3, 2.0)
P(Apartment | X) = 0.0000
P(House | X) = 0.7477
P(Condo | X) = 0.2523
Final Prediction (MAP) = House

Test Example 4: (1.5, 4, 1.5)
P(Apartment | X) = 1.0000
P(House | X) = 0.0000
P(Condo | X) = 0.0000
Final Prediction (MAP) = Apartment

Test Example 5: (1.5, 3, 2.0)
P(Apartment | X) = 0.0000
P(House | X) = 0.5684
P(Condo | X) = 0.4316
Final Prediction (MAP) = House


#Decision Tree Classifier

In [ ]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier, export_text
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# -----------------------------
# LOAD DATA
# -----------------------------
file = "Asssignment4_Data.xlsx"

train_df = pd.read_excel(file, sheet_name="Train")
test_df = pd.read_excel(file, sheet_name="Test")

# -----------------------------
# SPLIT FEATURES AND LABEL
# -----------------------------
X_train = train_df.drop(columns=["House ID", "Construction type"])
y_train = train_df["Construction type"]

X_test = test_df.drop(columns=["House ID", "Construction type"])
y_test = test_df["Construction type"]

# -----------------------------
# 1. DEFAULT DECISION TREE
# -----------------------------
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

train_pred = model.predict(X_train)
test_pred = model.predict(X_test)

train_acc = accuracy_score(y_train, train_pred)
test_acc = accuracy_score(y_test, test_pred)

print("DEFAULT TREE RESULTS")
print("Train Accuracy:", train_acc)
print("Test Accuracy:", test_acc)

# -----------------------------
# 2. MAX DEPTH EXPERIMENT
# -----------------------------
print("\nDEPTH EXPERIMENT RESULTS")

best_depth = 1
best_acc = 0

for d in range(1, 11):
    model = DecisionTreeClassifier(max_depth=d, random_state=42)
    model.fit(X_train, y_train)

    acc = accuracy_score(y_test, model.predict(X_test))
    print(f"Depth {d} -> Test Accuracy: {acc:.4f}")

    if acc > best_acc:
        best_acc = acc
        best_depth = d

print("\nBest Depth:", best_depth)
print("Best Accuracy:", best_acc)

# -----------------------------
# 3. FINAL TREE (BEST DEPTH)
# -----------------------------
final_model = DecisionTreeClassifier(max_depth=best_depth, random_state=42)
final_model.fit(X_train, y_train)

rules = export_text(final_model, feature_names=list(X_train.columns))
print("\nDECISION TREE RULES:\n")
print(rules)

DEFAULT TREE RESULTS
Train Accuracy: 1.0
Test Accuracy: 0.4

DEPTH EXPERIMENT RESULTS
Depth 1 -> Test Accuracy: 0.4000
Depth 2 -> Test Accuracy: 0.8000
Depth 3 -> Test Accuracy: 0.4000
Depth 4 -> Test Accuracy: 0.4000
Depth 5 -> Test Accuracy: 0.4000
Depth 6 -> Test Accuracy: 0.4000
Depth 7 -> Test Accuracy: 0.4000
Depth 8 -> Test Accuracy: 0.4000
Depth 9 -> Test Accuracy: 0.4000
Depth 10 -> Test Accuracy: 0.4000

Best Depth: 2
Best Accuracy: 0.8

DECISION TREE RULES:

|--- Age of home <= 36.00
|   |--- Local Price <= 8.41
|   |   |--- class: House
|   |--- Local Price >  8.41
|   |   |--- class: Apartment
|--- Age of home >  36.00
|   |--- Local Price <= 4.55
|   |   |--- class: Condo
|   |--- Local Price >  4.55
|   |   |--- class: Apartment



#Decision Tree Final Report

## 1. Default Decision Tree Results

### (a) Training Accuracy
The decision tree classifier was trained using default parameters from Scikit-learn.  
The model achieved a **training accuracy of 1.0**, meaning it perfectly classified all training samples.

This indicates that the model has fully fit the training data.

---

### (b) Test Accuracy
When evaluated on the test dataset, the model achieved a **test accuracy of 0.4**.  

This large gap between training and test accuracy suggests that the model is overfitting the training data and does not generalize well to unseen data.

---

## 2. Effect of Maximum Depth

Different values of maximum depth were tested from 1 to 10.

- At **depth = 1**, the model underfits and gives low accuracy (0.4).
- At **depth = 2**, the model achieves the best performance with a **test accuracy of 0.8**.
- For depths greater than 2, the accuracy decreases again to 0.4 due to overfitting.

### Conclusion:
Restricting the maximum depth improves generalization by reducing overfitting.  
The best depth for this dataset is **max_depth = 2**.

---

## 3. Decision Tree Rules (Max Depth = 2)
Age of home <= 36.00
|--- Local Price <= 8.41 → House
|--- Local Price > 8.41 → Apartment

Age of home > 36.00
|--- Local Price <= 4.55 → Condo
|--- Local Price > 4.55 → Apartment


---

## 4. Coverage and Accuracy of Rules

### Rule 1:
**IF Age of home ≤ 36 AND Local Price ≤ 8.41 → House**

- Coverage = 5 / 25 = 0.20  
- Accuracy = 4 / 5 = 0.80  

---

### Rule 2:
**IF Age of home ≤ 36 AND Local Price > 8.41 → Apartment**

- Coverage = 4 / 25 = 0.16  
- Accuracy = 3 / 4 = 0.75  

---

### Rule 3:
**IF Age of home > 36 AND Local Price ≤ 4.55 → Condo**

- Coverage = 3 / 25 = 0.12  
- Accuracy = 3 / 3 = 1.00  

---

### Rule 4:
**IF Age of home > 36 AND Local Price > 4.55 → Apartment**

- Coverage = 13 / 25 = 0.52  
- Accuracy = 10 / 13 ≈ 0.77  

---

## Final Conclusion

The decision tree performs best at **max depth = 2**, achieving the highest test accuracy (0.8).  
The most important features in classification are **Age of home** and **Local Price**.

Compared to the default model, restricting depth significantly reduces overfitting and improves generalization performance.